File: Chapter 03/Python/3.2 Coupling Exercise/Before/user_report_generator.py
from user_data_service import UserDataService

class UserReportGenerator:
    def __init__(self):
        self.data_service = UserDataService()

    def generate_report(self, user_id):
        name = self.data_service.get_user_name(user_id)
        email = self.data_service.get_user_email(user_id)
        orders = self.data_service.get_user_order_ids(user_id)

        total_spent = 0.0
        for order_id in orders:
            total_spent += self.data_service.get_order_total(order_id)

        return f"User Report for {name} ({email}) - Total Spent: ${total_spent:.2f}"

File: Chapter 03/C#/3.2 Coupling Exercise/Before/UserReportGenerator.cs
using System.Collections.Generic;

namespace Chapter03.CouplingTest.Before;

public class UserReportGenerator
{
    private readonly UserDataService _dataService = new();

    public string GenerateReport(int userId)
    {
        // High Coupling: The client has to orchestrate all the pieces
        string name = _dataService.GetUserName(userId);
        string email = _dataService.GetUserEmail(userId);
        List<string> orders = _dataService.GetUserOrderIds(userId); 

        decimal totalSpent = 0;
        foreach (var orderId in orders)
        {
            totalSpent += _dataService.GetOrderTotal(orderId);
        }

        return $"User Report for {name} ({email}) - Total Spent: ${totalSpent}";
    }
}